### PR TITLE
Introduce inactivity_timeout

### PR DESCRIPTION
**Pull request by TusharR-google**
#211

### CODE REVIEW

**Review by appsdesh — DISMISSED** [review]
Push and Pull are delivery mechanisms defined outside the SSF spec. We should not introduce new aspects specifically targeting a single delivery mechanism in the SSF spec

**Comment by FragLegs** [comment]
> Push and Pull are delivery mechanisms defined outside the SSF spec. We should not introduce new aspects specifically targeting a single delivery mechanism in the SSF spec

I agree. Can we just make this `stream_ttl` so that it applies to any stream regardless of delivery method?

**Comment by TusharR-google** [comment]
> Push and Pull are delivery mechanisms defined outside the SSF spec. We should not introduce new aspects specifically targeting a single delivery mechanism in the SSF spec

Thanks! The same concept can be applied to PULL streams too, let me reword this as simply stream_tll.

File: openid-sharedsignals-framework-1_0.md [hunk 1]
```diff
@@ -601,6 +618,13 @@ default_subjects
     to be transmitted. The Receiver MAY remove subjects added this way via the
     `remove_subject_endpoint`.
 
+push_stream_ttl
+
+> OPTIONAL. The lifetime of a PUSH stream in seconds, after which the Transmitter MAY either pause or disable the stream if it has not received any Receiver-initiated communication in that duration.
+  If the Transmitter decides to update the stream, it MUST send a Stream Updated Event to the Receiver as described in {{status}}.
+  If the Receiver calls any endpoint in the Event Stream Management API ({{management}}), the Transmitter must refresh the TTL of that particular stream.
```
Comment (FragLegs) [inline openid-sharedsignals-framework-1_0.md:625]:
```suggestion
  If the Receiver calls any endpoint in the Event Stream Management API ({{management}}), the Transmitter MUST refresh the TTL of that particular stream.
```

Comment (TusharR-google) [reply openid-sharedsignals-framework-1_0.md]: Done

Comment (FragLegs) [inline openid-sharedsignals-framework-1_0.md]: We might want to be explicit that contact from the receiver does not automatically move the stream out of a paused/disabled state caused by a TTL timeout.

File: openid-sharedsignals-framework-1_0.md [hunk 2]
```diff
@@ -601,6 +618,13 @@ default_subjects
     to be transmitted. The Receiver MAY remove subjects added this way via the
     `remove_subject_endpoint`.
 
+stream_ttl
+
+> OPTIONAL. The lifetime of a stream in seconds, after which the Transmitter MAY either pause or disable the stream if it has not received any Receiver-initiated communication in that duration.
```
Comment (tulshi) [inline openid-sharedsignals-framework-1_0.md:623]: "Receiver initiated" needs to be clarified, IMO. Does it differ for poll and push streams? Is the appropriate response code from the receiver to a push from a Tx sufficient?

File: openid-sharedsignals-framework-1_0.md [hunk 3]
```diff
@@ -601,6 +618,13 @@ default_subjects
     to be transmitted. The Receiver MAY remove subjects added this way via the
     `remove_subject_endpoint`.
 
+stream_ttl
```
Comment (ptravers) [inline openid-sharedsignals-framework-1_0.md:621]: can we make it `stream_ttl_seconds`? will save people working with the api a documentation lookup

File: openid-sharedsignals-framework-1_0.md (hunk 1 above)
Comment (TusharR-google) [reply openid-sharedsignals-framework-1_0.md]: Should we leave it up to the transmitter? It might complicate the spec if paused/disabled states were treated differently based on whether they were caused by a TTL timeout (Tx not allowed to automatically change the state), vs other causes like Rx requesting the pause (Tx allowed to do whatever).

File: openid-sharedsignals-framework-1_0.md (hunk 2 above)
Comment (TusharR-google) [reply openid-sharedsignals-framework-1_0.md]: Clarified the definition of "receiver-initiated" communication.

File: openid-sharedsignals-framework-1_0.md (hunk 3 above)
Comment (TusharR-google) [reply openid-sharedsignals-framework-1_0.md]: Most of the other names in the this/similar spec do not seem to use a suffix, regardless of units. E.g. min_verification_interval

So we could keep it as stream_ttl for consistency.

File: openid-sharedsignals-framework-1_0.md [hunk 4]
```diff
@@ -85,6 +91,17 @@ normative:
     target: https://tools.ietf.org/html/rfc6749#section-4.4
     title: The OAuth 2.0 Authorization Framework - Client Credentials Grant
 
+  EXPIRES_IN:
```
Comment (tulshi) [inline openid-sharedsignals-framework-1_0.md:94]: you don't need to describe an RFC. You can just add its number and the kramdown tool will do the rest.

File: openid-sharedsignals-framework-1_0.md [hunk 5]
```diff
@@ -929,6 +946,19 @@ description
   This is useful in multi-stream systems to identify the stream for human actors. The
   transmitter MAY truncate the string beyond an allowed max length.
 
+stream_ttl
+
+> **Transmitter-Supplied**, OPTIONAL. The refreshable lifetime of the stream in seconds, after which the Transmitter MAY either pause or disable the stream if it has not received any Receiver-initiated communication (defined below) in that duration. The syntax is the same as that of {{EXPIRES_IN}}.
```
Comment (tulshi) [inline openid-sharedsignals-framework-1_0.md:951]: You might want to expand the acronym "TTL" somewhere in the description. You say it precisely in the first sentence, but you just don't say what "TTL" stands for. It'll help clarify the name even though most folks will know what it is.

Comment (tulshi) [inline openid-sharedsignals-framework-1_0.md:951]: Is the right behavior "pause or disable", or just "delete"?

Comment (TusharR-google) [reply openid-sharedsignals-framework-1_0.md]: Pause / disable gives the Receiver an option to unpause / enable the stream, while delete means that the Receiver will have to recreate it.

I was imagining that some Transmitters may be more lenient where they simply pause the stream at first, then disable it after some time of being paused, and finally delete it if there's still no response (hence slowly escalating).

That said, either one works. Or we could even add the above paragraph as an example. What do you think?

Comment (TusharR-google) [reply openid-sharedsignals-framework-1_0.md]: SG, done.

File: openid-sharedsignals-framework-1_0.md (hunk 4 above)
Comment (TusharR-google) [reply openid-sharedsignals-framework-1_0.md:94]: I wasn't sure how to make it point to a specific appendix, hence followed the example of CLIENTCRED above. Both CLIENTCRED and EXPIRES_IN are from the same RFC, so the details remain the same.

Would you know the right syntax to make it point to the appendix instead of the whole RFC (https://tools.ietf.org/html/rfc6749#appendix-A.14).

File: openid-sharedsignals-framework-1_0.md (hunk 5 above)
Comment (FragLegs) [reply openid-sharedsignals-framework-1_0.md]: Since it is MAY, you could say "...after which the Transmitter MAY either pause, disable, or delete the stream if it has not received..."

Comment (TusharR-google) [reply openid-sharedsignals-framework-1_0.md]: Yeah we can keep the options open (implementation-dependent).

File: openid-sharedsignals-framework-1_0.md [hunk 6]
```diff
@@ -1123,6 +1140,19 @@ description
...
+> **Transmitter-Supplied**, OPTIONAL. The refreshable time-to-live (TTL) of the stream in seconds, after which the Transmitter MAY either pause or disable the stream if it has not received any Receiver-initiated communication (defined below) in that duration. The syntax is the same as that of {{EXPIRES_IN}}.
+>
+> For streams created with the PUSH {{RFC8935}} delivery method, the Transmitter MUST refresh the TTL whenever:
+> * The Receiver calls any endpoint in the Event Stream Management API ({{management}}).
+>
+> For streams created with the POLL {{RFC8936}} delivery method, the Transmitter MUST refresh the TTL whenever:
+> * The Receiver polls the Transmitter for events.
+> * The Receiver calls any endpoint in the Event Stream Management API ({{management}}).
+>
+> If the Transmitter decides to update the stream, it MUST send a Stream Updated Event to the Receiver as described in {{status}}.
```
Comment (FragLegs) [inline openid-sharedsignals-framework-1_0.md:1154]:
```suggestion
> If the Transmitter decides to pause or disable the stream, it MUST send a Stream Updated Event to the Receiver as described in {{status}}.
```
It took me a second to understand why this sentence was here. Specifying pause or disable connects it to the language above.

File: openid-sharedsignals-framework-1_0.md [hunk 7]
```diff
@@ -1123,6 +1140,19 @@ description
   This is useful in multi-stream systems to identify the stream for human actors. The
   transmitter MAY truncate the string beyond an allowed max length.
 
+inactivity_ttl
+
+> **Transmitter-Supplied**, OPTIONAL. The refreshable time-to-live (TTL) of the stream in seconds, after which the Transmitter MAY either pause, disable, or delete the stream if it has not received any Receiver-initiated communication (defined below) in that duration. The syntax is the same as that of {{EXPIRES_IN}}.
+>
+> For streams created with the PUSH {{RFC8935}} delivery method, the Transmitter MUST refresh the TTL whenever:
+> * The Receiver calls any endpoint in the Event Stream Management API ({{management}}).
```
Comment (tulshi) [inline openid-sharedsignals-framework-1_0.md:1148]: We probably need to specify here and below that "calls to the Event Stream Management API that reference the stream"

File: openid-sharedsignals-framework-1_0.md [hunk 8]
```diff
@@ -1123,6 +1140,19 @@ description
   This is useful in multi-stream systems to identify the stream for human actors. The
   transmitter MAY truncate the string beyond an allowed max length.
 
+inactivity_ttl
+
+> **Transmitter-Supplied**, OPTIONAL. The refreshable time-to-live (TTL) of the stream in seconds, after which the Transmitter MAY either pause, disable, or delete the stream if it has not received any Receiver-initiated communication (defined below) in that duration. The syntax is the same as that of {{EXPIRES_IN}}.
+>
+> For streams created with the PUSH {{RFC8935}} delivery method, the Transmitter MUST refresh the TTL whenever:
```
Comment (tulshi) [inline openid-sharedsignals-framework-1_0.md:1147]: "refresh the TTL" could be ambiguous. Can we be more precise? we could alter the description in line 1145 to read: "the refreshable time-to-live (TTL) of the stream in seconds, since the last activity from the Receiver as described below, was conducted. After the TTL, the Transmitter MAY..."

This sentence can then read: "Receiver activity can be any of:"

File: openid-sharedsignals-framework-1_0.md (hunk 4 above)
Comment (tulshi) [reply openid-sharedsignals-framework-1_0.md:100]: I guess you aren't really referring to "CLIENTCRED", right? The Expires_in definition in OAuth is pretty small, so one way you can do this is by referring to it as "Section A.14 of  RFC6749 {{RFC6749}}".

File: openid-sharedsignals-framework-1_0.md [hunk 9]
```diff
@@ -1123,6 +1140,19 @@ description
   This is useful in multi-stream systems to identify the stream for human actors. The
   transmitter MAY truncate the string beyond an allowed max length.
 
+inactivity_ttl
+
+> **Transmitter-Supplied**, OPTIONAL. The refreshable time-to-live (TTL) of the stream in seconds, after which the Transmitter MAY either pause, disable, or delete the stream if it has not received any Receiver-initiated communication (defined below) in that duration. The syntax is the same as that of {{EXPIRES_IN}}.
```
Comment (tulshi) [inline openid-sharedsignals-framework-1_0.md:1145]: see suggestion above.

File: openid-sharedsignals-framework-1_0.md (hunk 7 above)
Comment (TusharR-google) [reply openid-sharedsignals-framework-1_0.md]: Done

File: openid-sharedsignals-framework-1_0.md (hunk 9 above)
Comment (TusharR-google) [reply openid-sharedsignals-framework-1_0.md]: Done

**Review by tulshi — APPROVED** [review]

File: openid-sharedsignals-framework-1_0.md (hunk 4 above)
Comment (TusharR-google) [reply openid-sharedsignals-framework-1_0.md]: Done in a later commit as suggested

**Comment by TusharR-google** [comment]
> Push and Pull are delivery mechanisms defined outside the SSF spec. We should not introduce new aspects specifically targeting a single delivery mechanism in the SSF spec

#

**Review by FragLegs — APPROVED** [review]

File: openid-sharedsignals-framework-1_0.md [hunk 10]
```diff
@@ -1123,6 +1131,23 @@ description
   This is useful in multi-stream systems to identify the stream for human actors. The
   transmitter MAY truncate the string beyond an allowed max length.
 
+inactivity_ttl
+
+> **Transmitter-Supplied**, OPTIONAL. The refreshable time-to-live (TTL) of the stream in seconds, since the last activity from the Receiver as described below, was conducted. After the TTL passes with no eligible activity from the Receiver, the Transmitter MAY either pause, disable, or delete the stream. The syntax is the same as that of `expires_in` from Section A.14 of {{RFC6749}}.
```
Comment (FragLegs) [inline openid-sharedsignals-framework-1_0.md:1136]:
```suggestion
> **Transmitter-Supplied**, OPTIONAL. The refreshable time-to-live (TTL) of the stream in seconds. After the TTL passes with no eligible activity from the Receiver, as defined below, the Transmitter MAY either pause, disable, or delete the stream. The syntax is the same as that of `expires_in` from Section A.14 of {{RFC6749}}.
```
Nit suggestion for clarity

**Comment by tulshi** [comment]
> Push and Pull are delivery mechanisms defined outside the SSF spec. We should not introduce new aspects specifically targeting a single delivery mechanism in the SSF spec

Since the comments have been resolved in the PR, I'm going to "dismiss" this review. @appsdesh please comment if you think this still needs addressing.

File: openid-sharedsignals-framework-1_0.md [hunk 11]
```diff
@@ -1123,6 +1131,23 @@ description
...
+>
+> The following constitutes eligible Receiver activity, upon observing which, the Transmitter MUST reset its count toward the TTL of the stream to zero.
```
Comment (FragLegs) [inline openid-sharedsignals-framework-1_0.md:1138]:
```suggestion
> The following constitutes eligible Receiver activity. If the Transmitter observes any of these activities from the Receiver, it MUST restart the inactivity timeout counter.
```
Nit suggestion for clarity

**Review by FragLegs — APPROVED** [review]

**Review by tulshi — APPROVED** [review]

**Review by FragLegs — APPROVED** [review]
This looks great! Thanks for all the diligence getting it tuned up so well.

**Review by iamseanodentity — APPROVED** [review]

File: openid-sharedsignals-framework-1_0.md [hunk 12]
```diff
@@ -1123,6 +1131,23 @@ description
   This is useful in multi-stream systems to identify the stream for human actors. The
   transmitter MAY truncate the string beyond an allowed max length.
 
+inactivity_timeout
+
+> **Transmitter-Supplied**, OPTIONAL. The refreshable inactivity timeout of the stream in seconds. After the timeout duration passes with no eligible activity from the Receiver, as defined below, the Transmitter MAY either pause, disable, or delete the stream. The syntax is the same as that of `expires_in` from Section A.14 of {{RFC6749}}.
```
[the remark this answers is not in the document]
Comment (TusharR-google) [reply openid-sharedsignals-framework-1_0.md:1136]: For PUSH streams that's not enough, at least from a security/privacy perspective - receiver endpoints are more likely to be open, the only way to verify that the receiver exists and still has the right auth/access is for it to call the transmitter endpoints (oauth-protected in most cases).

For PULL, polling is sufficient.
